http-client-jdk/src/main/java/io/micronaut/http/client/jdk/ReactiveByteBufferByteBody.java

```diff
@@ -0,0 +1,508 @@
+        protected void addForward(List<ByteBufferConsumer> consumers) {
+            for (ByteBufferConsumer consumer : consumers) {
+                consumer.add(adding.asReadOnlyBuffer()); // we want independent positions
+            }
+        }
+
+        @Override
+        protected void addBuffer() {
+            if (buffer == null) {
+                buffer = new SnapshotByteArrayOutputStream();
+            }
+            buffer.write(adding);
+        }
+
+        @Override
+        protected void discardBuffer() {
+            buffer = null;
+        }
+
+        @Override
+        public void add(ByteBuffer buffer) {
+            submit(() -> {
+                adding = buffer;
+                add(buffer.remaining());
+                adding = null;
+            });
+        }
+
+        @Override
+        public void error(Throwable e) {
+            submit(() -> super.error(e));
+        }
+
+        @Override
+        public void complete() {
+            submit(super::complete);
+        }
+    }
+
+    /**
+     * {@link ByteArrayOutputStream} implementations that allows taking an efficient snapshot of
+     * the current data.
+     */
+    private static final class SnapshotByteArrayOutputStream extends ByteArrayOutputStream {
+        public ByteBuffer snapshot() {
+            return ByteBuffer.wrap(buf, 0, count).asReadOnlyBuffer();
+        }
+
+        public void write(ByteBuffer buffer) {
+            if (buffer.hasArray()) {
+                write(buffer.array(), buffer.arrayOffset() + buffer.position(), buffer.remaining());
+            } else {
+                byte[] b = new byte[buffer.remaining()];
+                buffer.get(buffer.position(), b);
+                write(b, 0, b.length);
+            }
+        }
+    }
+
+    private enum WorkState {
+        CLEAN,
+        WORKING_THEN_CLEAN,
+        WORKING_THEN_DIRTY
+    }
+
+    /**
+     * {@link HttpResponse.BodySubscriber} implementation that pushes data into a
+     * {@link SharedBuffer}.
+     */
+    static final class ByteBodySubscriber implements HttpResponse.BodySubscriber<CloseableByteBody>, BufferConsumer.Upstream {
+        private final SharedBuffer sharedBuffer;
+        private final CloseableByteBody root;
+        private final AtomicLong demand = new AtomicLong(0);
+        private Flow.Subscription subscription;
+        private boolean cancelled;
+        private volatile boolean disregardBackpressure;
+
+        ByteBodySubscriber(BodySizeLimits limits) {
+            sharedBuffer = new SharedBuffer(limits, this);
+            root = new ReactiveByteBufferByteBody(sharedBuffer);
+        }
+
+        @Override
+        public CompletionStage<CloseableByteBody> getBody() {
+            return CompletableFuture.completedFuture(root);
+        }
+
+        @Override
+        public void onSubscribe(Flow.Subscription subscription) {
+            boolean initialDemand;
+            boolean cancelled;
+            synchronized (this) {
```

Comment:
can these synchronised blocks be replaced by locks?

Reply:
it's not necessary because they only access fields, not call any other code